### ISSUE TITLE
Image Style Transfer Model not generating fake nodes

### ISSUE THREAD

**Issue by Shailendra2803** [tensorflow/tensorflow#24685]
HI,
I am trying to quantize the following model
https://github.com/hwalsuklee/tensorflow-fast-style-transfer

 I have trained the model with following 
tf.contrib.quantize.create_training_graph(quant_delay=0)
and for Inference I am using the following 
 tf.contrib.quantize.create_eval_graph()

I am freezing the graph with following command
python -m tensorflow.python.tools.freeze_graph \
--input_graph=FST_out.pb \
--input_checkpoint=q_out_trt1/final.ckpt-14000 \
--input_binary=true \
--output_graph=q_out_trt1/frozen_FST.pb \
--output_node_names=Tanh

What I have noticed after freezing the graph is that there is no Fake Quant Nodes .We have also noticed that most of the models which are generating fake nodes are slim based.
Please let us know the reason for not generating the fake nodes and if the quantization aware training works only for tf.contrib.slim based model.

Regards
Shailendra

**Comment by Shailendra2803** [comment]
Hi,

We have another observation that only when bias is added after convolution than only we are able to 
get the fake quantization layer.
In our network mentioned above we are not able to get the fake layer as there are no bias layer after Convolution layers.But if we add bias layers after convolution we are able to see fake layers.
Can you also please explain the behaviour.?

I request you to please check and  reply asap.

**Comment by msymp** [comment]
Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version. Also, did you compile from source or install a binary?

Make sure you also include the exact command if possible to produce the output included in your test case. If you are unclear what to include see the issue template displayed in [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new).

We ask for this in the issue submission template, because it is really difficult to help without that information. Thanks!
[New Issue Template](https://github.com/tensorflow/tensorflow/issues/new)

**Comment by Shailendra2803** [comment]
Hi ,
 System information
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- OS Platform and Distribution : Ubuntu 16.04
- TensorFlow installed from (source or binary):Binary
- TensorFlow version (use command below):tf1.10
- Python version:Python 2.7.14  and Python 3.6.3.

**Details :**
I am trying to quantize the following model
https://github.com/hwalsuklee/tensorflow-fast-style-transfer

I have trained the model with following
tf.contrib.quantize.create_training_graph(quant_delay=0)
and for Inference I am using the following
tf.contrib.quantize.create_eval_graph()

I am freezing the graph with following command
python -m tensorflow.python.tools.freeze_graph 
--input_graph=FST_out.pb 
--input_checkpoint=q_out_trt1/final.ckpt-14000 
--input_binary=true 
--output_graph=q_out_trt1/frozen_FST.pb 
--output_node_names=Tanh

What I have noticed after freezing the graph is that there is no Fake Quant Nodes .We have also noticed that most of the models which are generating fake nodes are slim based.
Please let us know the reason for not generating the fake nodes and if the quantization aware training works only for tf.contrib.slim based model.

We have another observation that only when bias is added after convolution than only we are able to
get the fake quantization layer.
In our network mentioned above we are not able to get the fake layer as there are no bias layer after Convolution layers.But if we add bias layers after convolution we are able to see fake layers.
Please look into the issue.

**How to train the model** : 
python run_train.py --style <style file> --output <output directory> --trainDB <trainDB directory> --vgg_model <model directory>

You need to download the Train DB from the link provided in the above mentioned link.

**How to take inference**
python run_test.py --content <content file> --style_model <style-model file> --output <output file> 

Regards
Shailendra Singh Yadav

**Comment by msymp** [comment]
Hi @petermattson , can you advise who would be able to address a fast style transfer query. Thanks.

**Comment by Shailendra2803** [comment]
Please also find the attached frozen file snippet generated using quantization aware training for further analysis
![frozen](https://user-images.githubusercontent.com/41136882/51240114-1092dc80-19a1-11e9-9322-5763d964075c.PNG)

 .



Regards
Shailendra

**Comment by petermattson** [comment]
Hi Shailendra -- sorry, I am completely unfamiliar with this issue. Maybe start with one of the slim primary contributors since nodes are slim based?

**Comment by msymp** [comment]
@Shailendra2803 , can you please communicate with the original contributors of Slim model:
https://github.com/tensorflow/models/tree/master/research/slim
Thanks.

**Comment by Shailendra2803** [comment]
@msymp  @petermattson 
Our model is not slim based.We are jst comparing with the slim based model for quantization aware training.
You can ignore all the slim based information provided here.The graph attached is for our model and its not slim based.
Main problem is that when we try to do quantization aware training for our model 
`def net(self, image):
        image_p = self._reflection_padding(image)
        conv1 = self._conv_layer(image_p, 32, 9, 1, name='conv1')
        conv2 = self._conv_layer(conv1, 64, 3, 2, name='conv2')
        conv3 = self._conv_layer(conv2, 128, 3, 2, name='conv3')
        resid1 = self._residual_block(conv3, 3, name='resid1')
        resid2 = self._residual_block(resid1, 3, name='resid2')
        resid3 = self._residual_block(resid2, 3, name='resid3')
        resid4 = self._residual_block(resid3, 3, name='resid4')
        resid5 = self._residual_block(resid4, 3, name='resid5')
        conv_t1 = self._conv_tranpose_layer(resid5, 64, 3, 2, name='convt1')
        conv_t2 = self._conv_tranpose_layer(conv_t1, 32, 3, 2, name='convt2')
        conv_t3 = self._conv_layer(conv_t2, 3, 9, 1, relu=False, name='convt3')
        preds = (tf.nn.tanh(conv_t3) + 1) * (255. / 2)
        return preds

    def _reflection_padding(self, net):
        return tf.pad(net,[[0, 0],[40, 40],[40, 40], [0, 0]], "REFLECT")

    def _conv_layer(self, net, num_filters, filter_size, strides, padding='SAME', relu=True, name=None):
        weights_init = self._conv_init_vars(net, num_filters, filter_size, name=name)
        strides_shape = [1, strides, strides, 1]
        net = tf.nn.conv2d(net, weights_init, strides_shape, padding=padding)
        net = self._instance_norm(net, name=name)
        if relu:
            net = tf.nn.relu(net)

        return net

    def _conv_tranpose_layer(self, net, num_filters, filter_size, strides, name=None):
        weights_init = self._conv_init_vars(net, num_filters, filter_size, transpose=True, name=name)

        batch_size, rows, cols, in_channels = [i.value for i in net.get_shape()]
        new_rows, new_cols = int(rows * strides), int(cols * strides)

        new_shape = [batch_size, new_rows, new_cols, num_filters]
        tf_shape = tf.stack(new_shape)
        strides_shape = [1,strides,strides,1]

        net = tf.nn.conv2d_transpose(net, weights_init, tf_shape, strides_shape, padding='SAME')
        net = self._instance_norm(net, name=name)
        return tf.nn.relu(net)

    def _residual_block(self, net, filter_size=3, name=None):
        batch, rows, cols, channels = [i.value for i in net.get_shape()]
        tmp = self._conv_layer(net, 128, filter_size, 1, padding='VALID', relu=True, name=name+'_1')
        return self._conv_layer(tmp, 128, filter_size, 1, padding='VALID', relu=False, name=name+'_2') + tf.slice(net, [0,2,2,0], [batch,rows-4,cols-4,channels])
`
We don't see any fake node generated.
In the above when we add a bias layer we can see fake node generated for all except the Transpose Conv.
Snippet of the code after bias layer is added is below
`    def _conv_layer(self, net, num_filters, filter_size, strides, padding='SAME', relu=True, name=None):
        weights_init = self._conv_init_vars(net, num_filters, filter_size, name=name)
        strides_shape = [1, strides, strides, 1]
        net = tf.nn.conv2d(net, weights_init, strides_shape, padding=padding)
        bias_init = self._bias_init_vars(net,num_filters,name=name)
        net = tf.nn.bias_add(net, bias_init)
        net = self._instance_norm(net, name=name)
        if relu:
            net = tf.nn.relu(net)

        return net
`

So can you please provide input for the following
1.)Why fake node is not generated for all the layers without addition of bias layer ?
2.)Why fake node is not generated for Transpose conv even when bias layer is added?

**Comment by ymodak** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new). Thanks!

**Comment by SystemErrorWang** [comment]
@Shailendra2803 i don't know if quantization only works for slim, but there seems to be convolution_transpose layers in style transfer model, which is not yet supported in tf lite quantize aware training.
